src/types/routeMethods.ts

```diff
@@ -58,20 +58,20 @@ export type ExtractRouteMethodParams<T> = T extends RouteMethod<infer Params>
 export type ExtractParamsFromPath<
   TPath extends Route['path']
 > = TPath extends Path
-  ? ExtractParamsFromPathString<UnifyParamEnds<TPath['path']>, TPath['params']>
+  ? TPath['params']
   : TPath extends string
-    ? ExtractParamsFromPathString<UnifyParamEnds<TPath>>
+    ? Path<TPath, {}>['params']
     : never
 
 type ParamEnd = '/'
 
-type UnifyParamEnds<
+export type UnifyParamEnds<
   TPath extends string
 > = ReplaceAll<ReplaceAll<TPath, '-', ParamEnd>, '_', ParamEnd>
 
 export type ExtractParamsFromPathString<
   TPath extends string,
-  TParams extends Record<string, Param> = Record<never, never>
+  TParams extends Record<string, Param | undefined> = Record<never, never>
```

Comment:
These `| undefined` additions just account for the user defined param types being an optional record which adds `| undefined` to every property.